libbpf-tools/tcpretrans.bpf.c

```diff
@@ -0,0 +1,150 @@
+#define AF_INET		2
+#define AF_INET6	10
+
+const volatile bool do_count = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv4_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv4_count SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv6_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv6_count SEC(".maps");
+
+static void count_v4(const struct sock *skp)
+{
+	struct ipv4_flow_key key = {};
+	static __u64 zero;
+	__u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, skp, __sk_common.skc_rcv_saddr);
+	BPF_CORE_READ_INTO(&key.daddr, skp, __sk_common.skc_daddr);
+	BPF_CORE_READ_INTO(&key.dport, skp, __sk_common.skc_dport);
+	BPF_CORE_READ_INTO(&key.sport, skp, __sk_common.skc_num);
+	val = bpf_map_lookup_or_try_init(&ipv4_count, &key, &zero);
+	if (val)
+		__atomic_add_fetch(val, 1, __ATOMIC_RELAXED);
+}
+
+static void count_v6(const struct sock *skp)
+{
+	struct ipv6_flow_key key = {};
+	static const __u64 zero;
+	__u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, skp,
+			   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&key.daddr, skp,
+			   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&key.dport, skp, __sk_common.skc_dport);
+	BPF_CORE_READ_INTO(&key.sport, skp, __sk_common.skc_num);
+
+	val = bpf_map_lookup_or_try_init(&ipv6_count, &key, &zero);
+	if (val)
+		__atomic_add_fetch(val, 1, __ATOMIC_RELAXED);
+}
+
+static int trace_event(void *ctx, const struct sock *skp, int type)
+{
+	if (skp == NULL)
+		return 0;
+
+	struct event e = {};
+	__u32 family;
+	__u64 pid_tgid;
+	__u32 pid;
+	int state;
+
+	family = BPF_CORE_READ(skp, __sk_common.skc_family);
+	e.af = family;
+
+	if (do_count) {
+		if (family == AF_INET)
+			count_v4(skp);
+		else
+			count_v6(skp);
+		return 0;
+	}
+
+	e.type = type;
+	pid_tgid = bpf_get_current_pid_tgid();
+	pid = pid_tgid >> 32;
+	e.pid = pid;
+
+	BPF_CORE_READ_INTO(&e.dport, skp, __sk_common.skc_dport);
+	BPF_CORE_READ_INTO(&e.sport, skp, __sk_common.skc_num);
+	state = BPF_CORE_READ(skp, __sk_common.skc_state);
+	e.state = state;
+
+	if (family == AF_INET) {
+		BPF_CORE_READ_INTO(&e.saddr, skp, __sk_common.skc_rcv_saddr);
+		BPF_CORE_READ_INTO(&e.daddr, skp, __sk_common.skc_daddr);
+	} else if (family == AF_INET6) {
+		BPF_CORE_READ_INTO(&e.saddr, skp,
+				   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+		BPF_CORE_READ_INTO(&e.daddr, skp,
+				   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	}
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+				  &e, sizeof(e));
+	return 0;
+}
+
+SEC("tp/tcp/tcp_retransmit_skb")
+int tracepoint__tcp__tcp_retransmit_skb(struct trace_event_raw_tcp_event_sk_skb* ctx)
```

Comment:
this is super long and inconvenient name, it will also get truncated to generic "tracepoint_" or something like that, if you dump maps with bpftool. We still have few original tools using such BCC-style long names (where it's actually part of defining which tracepoint to connect too), but generally we've moved to shorter names, like  just "tcp_retransmit_skb" would be totally appropriate in this case.